File: backend.core/pkg/bible/usx/metadata.go
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package usx

type MetadataIdentification struct {
	Name              string `xml:"name"`
	NameLocal         string `xml:"nameLocal"`
	Description       string `xml:"description"`
	Abbreviation      string `xml:"abbreviation"`
	AbbreviationLocal string `xml:"abbreviationLocal"`
	Scope             string `xml:"scope"`
	DateCompleted     string `xml:"dateCompleted"`
	BundleProducer    string `xml:"bundleProducer"`
	SystemId          []struct {
		Type     string `xml:"type,attr"`
		ID       string `xml:"id"`
		CsetId   string `xml:"csetId"`
		Name     string `xml:"name"`
		FullName string `xml:"fullName"`
	} `xml:"systemId"`
}

type MetadataType struct {
	Medium          string `xml:"medium"`
	IsConfidential  string `xml:"isConfidential"`
	HasCharacters   string `xml:"hasCharacters"`
	IsTranslation   string `xml:"isTranslation"`
	IsExpression    string `xml:"isExpression"`
	TranslationType string `xml:"translationType"`
	Audience        string `xml:"audience"`
	ProjectType     string `xml:"projectType"`
}

type MetadataLanguage struct {
	Iso             string `xml:"iso"`
	Name            string `xml:"name"`
	NameLocal       string `xml:"nameLocal"`
	Script          string `xml:"script"`
	ScriptCode      string `xml:"scriptCode"`
	ScriptDirection string `xml:"scriptDirection"`
	Ldml            string `xml:"ldml"`
	Rod             string `xml:"rod"`
	Numerals        string `xml:"numerals"`
}

type MetadataName struct {
	ID    string `xml:"id,attr"`
	Abbr  string `xml:"abbr"`
	Short string `xml:"short"`
	Long  string `xml:"long"`
}

type MetadataRightsHolder struct {
	Abbr      string `xml:"abbr"`
	URL       string `xml:"url"`
	NameLocal string `xml:"nameLocal"`
	Uid       string `xml:"uid"`
	Name      string `xml:"name"`
}

type MetadataRightsAdmin struct {
	URL  string `xml:"url"`
	Uid  string `xml:"uid"`
	Name string `xml:"name"`
}

type MetadataContributor struct {
	Content     string `xml:"content"`
	Publication string `xml:"publication"`
	Management  string `xml:"management"`
	Finance     string `xml:"finance"`
	Qa          string `xml:"qa"`
	Uid         string `xml:"uid"`
	Name        string `xml:"name"`
}

type MetadataAgencies struct {
	RightsHolder MetadataRightsHolder `xml:"rightsHolder"`
	RightsAdmin  MetadataRightsAdmin  `xml:"rightsAdmin"`
	Contributor  MetadataContributor  `xml:"contributor"`
}

type MetadataCountry struct {
	Iso  string `xml:"iso"`
	Name string `xml:"name"`
}

type MetadataCountries struct {
	Country MetadataCountry `xml:"country"`
}

type MetadataBook struct {
	Code string `xml:"code,attr"`
}

type MetadataResource struct {
	Checksum string `xml:"checksum,attr"`
	MimeType string `xml:"mimeType,attr"`
	Size     string `xml:"size,attr"`
	URI      string `xml:"uri,attr"`
}

type MetadataStructureContent struct {
	Name string `xml:"name,attr"`
	Src  string `xml:"src,attr"`
	Role string `xml:"role,attr"`
}

type MetadataPublication struct {
	ID                string `xml:"id,attr"`
	Default           string `xml:"default,attr"`
	Name              string `xml:"name"`
	NameLocal         string `xml:"nameLocal"`
	Description       string `xml:"description"`
	DescriptionLocal  string `xml:"descriptionLocal"`
	Abbreviation      string `xml:"abbreviation"`
	AbbreviationLocal string `xml:"abbreviationLocal"`
	CanonicalContent  struct {
		Book []MetadataBook `xml:"book"`
	} `xml:"canonicalContent"`
	Structure struct {
		Content []MetadataStructureContent `xml:"content"`
	} `xml:"structure"`
}

type Metadata struct {
	ID             string                 `xml:"id,attr"`
	Version        string                 `xml:"version,attr"`
	Revision       string                 `xml:"revision,attr"`
	Identification MetadataIdentification `xml:"identification"`
	Type           MetadataType           `xml:"type"`
	Relationships  string                 `xml:"relationships"`
	Agencies       MetadataAgencies       `xml:"agencies"`
	Language       MetadataLanguage       `xml:"language"`
	Countries      MetadataCountries      `xml:"countries"`
	Format         struct {
		UsxVersion       string `xml:"usxVersion"`
		VersedParagraphs string `xml:"versedParagraphs"`
	} `xml:"format"`
	Names struct {
		Name []MetadataName `xml:"name"`
	} `xml:"names"`
	Manifest struct {
		Resource []MetadataResource `xml:"resource"`
	} `xml:"manifest"`
	Source struct {
		CanonicalContent struct {
			Book []MetadataBook `xml:"book"`
		} `xml:"canonicalContent"`
		Structure struct {
			Content struct {
				Src  string `xml:"src,attr"`
				Role string `xml:"role,attr"`
			} `xml:"content"`
		} `xml:"structure"`
	} `xml:"source"`
	Publications struct {
		Publication MetadataPublication `xml:"publication"`
	} `xml:"publications"`
	Copyright struct {
		Text          string `xml:",chardata"`
		FullStatement struct {
			Text             string `xml:",chardata"`
			StatementContent struct {
				Text string `xml:",chardata"`
				Type string `xml:"type,attr"`
				P    struct {
					Text   string `xml:",chardata"`
					Strong string `xml:"strong"`
				} `xml:"p"`
			} `xml:"statementContent"`
		} `xml:"fullStatement"`
	} `xml:"copyright"`
	Promotion struct {
		PromoVersionInfo struct {
			Text        string `xml:",chardata"`
			ContentType string `xml:"contentType,attr"`
			P           []struct {
				Text string `xml:",chardata"`
				A    struct {
					Text string `xml:",chardata"`
					Href string `xml:"href,attr"`
				} `xml:"a"`
			} `xml:"p"`
		} `xml:"promoVersionInfo"`
	} `xml:"promotion"`
	ArchiveStatus struct {
		Text          string `xml:",chardata"`
		ArchivistName string `xml:"archivistName"`
		DateArchived  string `xml:"dateArchived"`
		DateUpdated   string `xml:"dateUpdated"`
		Comments      string `xml:"comments"`
	} `xml:"archiveStatus"`
}

File: workspace/bible/processor/src/metadata.ts
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export class ManifestName {
  private name: any

  constructor(name: any) {
    this.name = name
  }

  get id(): string {
    return this.name.$.id
  }

  get abbreviation(): string {
    const name = this.name.abbr
    if (Array.isArray(name)) return name[0]
    return name
  }

  get short(): string {
    const name = this.name.short
    if (Array.isArray(name)) return name[0]
    return name
  }

  get long(): string {
    const name = this.name.long
    if (Array.isArray(name)) return name[0]
    return name
  }
}

export class MetadataSystemId {
  private systemId: any

  constructor(systemId: any) {
    this.systemId = systemId
  }

  get id(): string {
    for (const id of this.systemId) {
      if (id.$.type === 'paratext') {
        return id.id[0]
      }
    }
    throw new Error('unknown id')
  }
}

export class MetadataIdentification {
  private identification: any
  readonly systemId: MetadataSystemId

  constructor(identification: any) {
    this.identification = identification
    this.systemId = new MetadataSystemId(identification.systemId)
  }

  get name(): string {
    return this.identification.name[0]
  }

  get nameLocal(): string {
    return this.identification.nameLocal[0]
  }

  get description(): string {
    return this.identification.description[0]
  }

  get abbreviation(): string {
    return this.identification.abbreviation[0]
  }

  get abbreviationLocal(): string {
    return this.identification.abbreviationLocal[0]
  }
}

export class MetadataLanguage {
  private language: any

  constructor(language: any) {
    this.language = language
  }

  get iso(): string {
    return this.language.iso[0]
  }

  get name(): string {
    return this.language.name[0]
  }

  get nameLocal(): string {
    return this.language.nameLocal[0]
  }

  get script(): string {
    return this.language.script[0]
  }

  get scriptCode(): string {
    return this.language.scriptCode[0]
  }

  get scriptDirection(): string {
    return this.language.scriptDirection[0]
  }
}

export class Publication {
  readonly names: ManifestName[]
  readonly contents: { [id: string]: PublicationContent } = {}

  constructor(names: ManifestName[], publication: any) {
    this.names = names
    for (const content of publication.structure[0].content) {
      const p = new PublicationContent(content)
      this.contents[p.id] = p
    }
  }
}

export class PublicationContent {
  private readonly content: any

  constructor(content: any) {
    this.content = content
  }

  get id(): string {
    return this.content.$.name
  }

  get file(): string {
    return this.content.$.src
  }

  get usfm(): string {
    return this.content.$.role
  }
}

export class BibleMetadata {
  readonly identification: MetadataIdentification
  readonly language: MetadataLanguage
  readonly publication!: Publication

  constructor(metadata: any) {
    this.identification = new MetadataIdentification(metadata.identification[0])
    this.language = new MetadataLanguage(metadata.language[0])
    const names: ManifestName[] = []
    for (const name of metadata.names[0].name) {
      names.push(new ManifestName(name))
    }
    this.publication = new Publication(names, metadata.publications[0].publication[0])
  }
}